File: go/src/shopping_cart.go
package shoppingcart

import "errors"

type Stock struct {
	Count   int
	Product *Product
}

type Order struct {
	Count int
	Total int
}

type ShoppingCart struct {
	Products map[string]Stock
}

func (shoppingCart *ShoppingCart) Add(code string, count int) int {
	v, exists := shoppingCart.Products[code]
	if exists {
		v.Count += count
		shoppingCart.Products[code] = v
		return v.Count
	} else {
		shoppingCart.Products[code] = Stock{Count: count, Product: &Product{code: code}}
		return count
	}
}

func (shoppingCart *ShoppingCart) Remove(code string, count int) int {
	v, exists := shoppingCart.Products[code]
	if exists {
		v.Count -= count
		shoppingCart.Products[code] = v
		// 0個になったらmapから削除
		if v.Count == 0 {
			delete(shoppingCart.Products, code)
			return 0
		}
		return v.Count
	} else {
		return 0
	}
}

func (shoppingCart *ShoppingCart) Order() (*Order, error) {
	if len(shoppingCart.Products) == 0 {
		return nil, errors.New("カートが空なので注文できません")
	}

	count := 0
	total := 0
	for _, stock := range shoppingCart.Products {
		price, err := stock.Product.Price()
		if err != nil {
			return nil, err
		}
		count += stock.Count
		total += stock.Count * price
	}

	return &Order{
		Count: count,
		Total: total,
	}, nil
}

File: typescript/src/shoppingCart.ts
import { Product } from "./product";

export type Stock = {
  product: Product;
  count: number;
};

export type Order = { count: number; total: number };

export class ShoppingCart {
  /**
   * カートに入った商品情報
   * @readonly
   * @var {{[key: string]: Stock}}
   */
  public readonly products: {
    [key: string]: Stock;
  } = {};

  /**
   * カートに商品を追加する
   * @param {string} code
   * @param {number} count
   * @returns {number} 追加後の個数を返却する
   */
  public add(code: string, count: number = 1): number {
    if (code in this.products) {
      this.products[code].count += count;
    } else {
      this.products[code] = {
        product: new Product(code),
        count,
      };
    }
    return this.products[code].count;
  }

  /**
   * カートから商品を取り出す
   * @param code
   * @param count
   * @returns {number} 削除後の個数を返却する
   */
  public remove(code: string, count: number = 1): number {
    if (code in this.products) {
      this.products[code].count -= count;
      // 個数が0になったら削除する
      if (this.products[code].count === 0) {
        delete this.products[code];
        return 0;
      } else {
        return this.products[code].count;
      }
    } else {
      return 0;
    }
  }

  /**
   * 注文する
   * @returns {Order} 商品点数と合計金額を返却する
   */
  public order(): Order {
    if (Object.keys(this.products).length === 0) {
      throw new Error("カートが空なので注文できません");
    }

    const order = { count: 0, total: 0 };
    for (const { count, product } of Object.values(this.products)) {
      order.count += count;
      order.total += product.price * count;
    }
    return order;
  }
}
